konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineProfiler.java

```diff
@@ -0,0 +1,204 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+package ai.konduit.serving.pipeline.api.context;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.FilenameUtils;
+import org.nd4j.common.io.StringUtils;
+import org.nd4j.common.primitives.AtomicBoolean;
+import org.nd4j.shade.jackson.databind.ObjectMapper;
+
+import java.io.*;
+import java.lang.management.ManagementFactory;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+
+@Slf4j
+public class PipelineProfiler implements Profiler {
+
+    private Writer writer;
+    private ObjectMapper json;
+    private final Thread fileWritingThread;
+    private final BlockingQueue<TraceEvent> writeQueue;
+    private final AtomicBoolean writing = new AtomicBoolean(false);
+
+    private long startTime;
+    private long endTime;
+    private final long pid;
+    private final long tid;
+
+    @Getter
+    private boolean logActive;
+    private ProfilerConfig profilerConfig;
+
+    private long getProcessId() {
+        // Note: may fail in some JVM implementations
+        // therefore fallback has to be provided
+
+        // something like '<pid>@<hostname>', at least in SUN / Oracle JVMs
+        final String jvmName = ManagementFactory.getRuntimeMXBean().getName();
+        final int index = jvmName.indexOf('@');
+
+        if (index < 1) {
+            // part before '@' empty (index = 0) / '@' not found (index = -1)
+            return 0;
+        }
+
+        try {
+            return Long.parseLong(jvmName.substring(0, index));
+        } catch (NumberFormatException e) {
+            // ignore
+        }
+        return 0;
+    }
+
+    private void fileSizeGuard() throws IOException {
+        System.out.println("Size is " + Files.size(profilerConfig.getOutputFile()));
+        if (Files.size(profilerConfig.getOutputFile()) > profilerConfig.getSplitSize()) {
+            Path currentFile = profilerConfig.getOutputFile();
+            String baseName = FilenameUtils.removeExtension(currentFile.getFileName().toString());
+            /*String parts[] = baseName.split("_");
+            int num = 1;
+            if (parts.length > 1) {
+                num = Integer.parseInt(parts[1]) + 1;
+            }*/
+            Path backup = Paths.get(currentFile.getParent() + FileSystems.getDefault().getSeparator() +
+                    baseName + "_" + String.valueOf(System.currentTimeMillis()) + ".json");
+            Files.move(currentFile, backup);
+            Files.createFile(currentFile);
+        }
+    }
```

Comment:
We should only call fileSizeGuard() from the writing thread, to avoid race conditions.
i.e., it might (try to) move the file right when the thread is in the middle of writing. Plus given we're using a buffered writer, we should be closing (or at very least flushing) the writer before we do anything to the file.